### PR TITLE
Allow better subclassing of MoneyField

### PR DESCRIPTION
**Pull request by beerbohmdo**
## Description

Move generation of NumberField from constructor to method to allow override in subclass.

## Issues

- https://github.com/silverstripe/silverstripe-framework/issues/11153

## Pull request checklist

- [x] The target branch is correct
    - See [picking the right version](https://docs.silverstripe.org/en/contributing/code/#picking-the-right-version)
- [x] All commits are relevant to the purpose of the PR (e.g. no debug statements, unrelated refactoring, or arbitrary linting)
    - Small amounts of additional linting are usually okay, but if it makes it hard to concentrate on the relevant changes, ask for the unrelated changes to be reverted, and submitted as a separate PR.
- [x] The commit messages follow our [commit message guidelines](https://docs.silverstripe.org/en/contributing/code/#commit-messages)
- [x] The PR follows our [contribution guidelines](https://docs.silverstripe.org/en/contributing/code/)
- [x] Code changes follow our [coding conventions](https://docs.silverstripe.org/en/contributing/coding_conventions/)
- [x] This change is covered with tests (or tests aren't necessary for this change)
- [ ] Any relevant User Help/Developer documentation is updated; for impactful changes, information is added to the changelog for the intended release
- [x] CI is green

### CODE REVIEW

File: src/Forms/MoneyField.php [hunk 1]
```diff
@@ -75,6 +71,18 @@ public function __clone()
         $this->fieldCurrency = clone $this->fieldCurrency;
     }
 
+    /**
+     * @return void
```
Comment (GuySartorelli) [inline src/Forms/MoneyField.php]:
```suggestion
     Builds a field to input the amount of money
```

File: src/Forms/MoneyField.php [hunk 2]
```diff
@@ -75,6 +71,18 @@ public function __clone()
...
+     */
+    protected function buildAmountField()
```
Comment (GuySartorelli) [inline src/Forms/MoneyField.php]:
```suggestion
    protected function buildAmountField(): void
```

**Review by GuySartorelli — CHANGES_REQUESTED** [review]
Seems sensible. Thanks for contributing!

Couple of small changes.
If there's an existing test for `buildCurrencyField()`, please add one for the new method as well.

## your pr description
Please fill in all sections of the template, including linking to an issue.
If no issue currently exists, please create one. We track issues, not pull requests.

Please also tick all of the check boxes that apply - if you find that you can't tick a checkbox it probably means there's some more work for you to do (e.g. the commit message doesn't match the guidelines right now)

**Comment by beerbohmdo** [comment]
> If there's an existing test for buildCurrencyField(), please add one for the new method as well.

There was no test for this method, I am not even sure what you wan't to test here. Nevertheless, I've added two test methods.


I will create a issue for this.

File: src/Forms/MoneyField.php (hunk 1 above)
Comment (GuySartorelli) [reply src/Forms/MoneyField.php]: Still need to remove this per the above review
```suggestion
```

File: tests/php/Forms/MoneyFieldTest.php [hunk 3]
```diff
@@ -127,4 +131,31 @@ public function testValidation()
         ]);
         $this->assertFalse($field->validate($validator));
     }
+
+    public function testCurrencyField(): void
```
Comment (GuySartorelli) [inline tests/php/Forms/MoneyFieldTest.php]:
```suggestion
    public function testGetCurrencyField(): void
```
Where possible, we try to name test methods `test<MethodName>` so it's immediately clear that it's testing a specific method on the class.

File: tests/php/Forms/MoneyFieldTest.php [hunk 4]
```diff
@@ -127,4 +131,31 @@ public function testValidation()
...
+        $this->assertInstanceOf(HiddenField::class, $field->getCurrencyField());
+        $this->assertEquals('Money[Currency]', $field->getCurrencyField()->getName());
+
+        $field->setAllowedCurrencies([]);
+
+        $this->assertInstanceOf(TextField::class, $field->getCurrencyField());
+        $this->assertEquals('Money[Currency]', $field->getCurrencyField()->getName());
+    }
+
+    public function testAmountField(): void
```
Comment (GuySartorelli) [inline tests/php/Forms/MoneyFieldTest.php]:
```suggestion
    public function testGetAmountField(): void
```

**Review by GuySartorelli — CHANGES_REQUESTED** [review]
Thank you for the tests, that's fantastic! Exactly what I was after.

Just a couple of small things to tidy up, then I'll be happy to merge this.

**Review by GuySartorelli — APPROVED** [review]
Fantastic, thank you for implementing this and for your patience.

**Pull request merged**